src/cpp/fastdds/core/policy/QosPoliciesSerializer.hpp

```diff
@@ -1240,6 +1245,645 @@ inline bool QosPoliciesSerializer<TopicDataQosPolicy>::read_from_cdr_message(
+    if (parameter_length < PARAMETER_BOOL_LENGTH)
+    {
+        return false;
+    }
+
+    bool valid = rtps::CDRMessage::readOctet(cdr_message,
+                    (fastdds::rtps::octet*)&qos_policy.autodispose_unregistered_instances);
+    cdr_message->pos += 3; //padding
+    return valid;
+}
+
+template<>
+inline uint32_t QosPoliciesSerializer<PublishModeQosPolicy>::cdr_serialized_size(
+        const PublishModeQosPolicy& qos_policy)
+{
+    // p_id + p_length + kind(1) + padding(3)
+    uint32_t ret_val = 2 + 2 + 1 + 3;
+    // + str_size + str_data (including null char)
+    ret_val += 4 + static_cast<uint32_t>(qos_policy.flow_controller_name.size());
+    // align
+    ret_val = (ret_val + 3) & ~3;
+
+    return ret_val;
+}
+
+template<>
+inline bool QosPoliciesSerializer<PublishModeQosPolicy>::add_to_cdr_message(
+        const PublishModeQosPolicy& qos_policy,
+        rtps::CDRMessage_t* cdr_message)
+{
+    bool valid = rtps::CDRMessage::addOctet(cdr_message, qos_policy.kind);
+    valid &= rtps::CDRMessage::addOctet(cdr_message, 0); // padding
+    valid &= rtps::CDRMessage::addOctet(cdr_message, 0); // padding
+    valid &= rtps::CDRMessage::addOctet(cdr_message, 0); // padding
+
+    valid &= rtps::CDRMessage::add_string(cdr_message, qos_policy.flow_controller_name);
+
+    return valid;
+}
+
+template<>
+inline bool QosPoliciesSerializer<PublishModeQosPolicy>::read_content_from_cdr_message(
+        PublishModeQosPolicy& qos_policy,
+        rtps::CDRMessage_t* cdr_message,
+        const uint16_t parameter_length)
+{
+    // Fail if length is lower than kind + str_size + (null str + padding)
+    if (parameter_length < (PARAMETER_KIND_LENGTH + 4 + 4))
+    {
+        return false;
+    }
+
+    uint32_t pos_ref = cdr_message->pos;
+
+    bool valid = rtps::CDRMessage::readOctet(cdr_message,
+                    (fastdds::rtps::octet*)&qos_policy.kind);
+    cdr_message->pos += 3; //padding
+
+    rtps::CDRMessage::readString(cdr_message, &qos_policy.flow_controller_name);
+
+    uint32_t length_diff = cdr_message->pos - pos_ref;
+    valid &= (parameter_length == length_diff);
+
+    return valid;
+}
+
+template<>
+inline uint32_t QosPoliciesSerializer<RTPSReliableWriterQos>::cdr_serialized_size(
+        const RTPSReliableWriterQos&)
+{
+    // p_id + p_length + times(32)
+    uint32_t ret_val = 2 + 2 + 32;
+
+    // + disable_positive_acks(12)
+    ret_val += 12;
+
+    // + disable_heatbeat_piggyback(4)
+    ret_val += 4;
+
+    return ret_val;
+}
+
+template<>
+inline bool QosPoliciesSerializer<RTPSReliableWriterQos>::add_to_cdr_message(
+        const RTPSReliableWriterQos& qos_policy,
+        rtps::CDRMessage_t* cdr_message)
+{
+    // times
+    bool valid = rtps::CDRMessage::addInt32(cdr_message,
+                    qos_policy.times.initial_heartbeat_delay.seconds);
+    valid &= rtps::CDRMessage::addUInt32(cdr_message,
+                    qos_policy.times.initial_heartbeat_delay.fraction());
```

Comment:
I would use nanoseconds here, since it is what the user configures

Reply:
Applies elsewhere

Reply:
We could have methods for Duration de/serialization

Reply:
I am only applying the duration factor-out in the methods added in this PR since other serializations (lifespan for instance) is serializing the fraction instead

Reply:
Yes. This only applies to custom PIDs.